FHEM/lib/SD_ProtocolData.pm

```diff
@@ -3297,6 +3297,22 @@ package lib::SD_ProtocolData;
         clientmodule    => 'SD_WS',
         length_min      => '36',
       },
+    "125" =>  ## Humidity and Temperaturesensor Ecowitt WH31,  MN;D=3002826E3681D96C000200000043;R=56;
+      {
+        name            => 'WH31',
+        comment         => 'Fine Offset | Ambient Weather WH31E Thermo-Hygrometer Sensor',
+        id              => '125',
+        knownFreqs      => '868.35',
+        datarate        => '17.257',
+        sync            => '2DD4',
+        modulation      => '2-FSK',
+        regexMatch      => qr/^(30|37)/, 
+        preamble        => 'W125#',
+        register        => ['0001','022E','0343','042D','05D4','0609','0780','0800','0D21','0E65','0FE8','10A9','115C','1202','1322','14F8','1543','1916','1B43','1C68'],
+        rfmode          => 'Fine_Offset_WH57_868',
+        clientmodule    => 'SD_WS',
+        length_min      => '28',
+      },
```

Comment:
Ich würde besser einen neuen rfmode definieren, da die Anzahl der Bytes nicht passt:

```
    "125" =>  ## Humidity and Temperaturesensor Ecowitt WH31, froggit DP50 / WH31A
              # Nordamerika: 915MHz; Europa: 868MHz, andere Regionen: 433MHz
              # https://github.com/RFD-FHEM/RFFHEM/pull/1161 @ sidey79 2023-04-01
              # MN;D=3002826E3681D96C000200000043;R=56;
      {
        name            => 'WH31',
        comment         => 'Fine Offset | Ambient Weather WH31E Thermo-Hygrometer Sensor',
        id              => '125',
        knownFreqs      => '868.35',
        datarate        => '17.257',
        sync            => '2DD4',
        modulation      => '2-FSK',
        regexMatch      => qr/^(30|37)/, 
        preamble        => 'W125#',
        # change register 0x03 FIFOTHR '0343' -> '0341' (8 byte), 0x06 PKTLEN '0609' -> '0607' (7 byte)
        register        => ['0001','022E','0341','042D','05D4','0607','0780','0800','0D21','0E65','0FE8','10A9','115C','1202','1322','14F8','1543','1916','1B43','1C68'],
        # register        => ['0001','022E','0343','042D','05D4','0609','0780','0800','0D21','0E65','0FE8','10A9','115C','1202','1322','14F8','1543','1916','1B43','1C68'],
        rfmode          => 'Fine_Offset_WH31_868',
        clientmodule    => 'SD_WS',
        length_min      => '14',
      },

```

Reply:
Also vom Prinzip bräuchten wir einen RF Mode, der alle Sensoren aus dieser Serie unterstützt, denn wir möchten ja alle empfangen können.

Reply:
Naja, das geht im Prinzip, wenn man den rfmode mit der größten Paketlänge einstellt. So empfange ich z..B. mit rfmode Bresser_5in1 die Protokolle 108, 115 und 117.

Das hat allerdings den Nachteil, das der CC1101 bei kürzeren Protokollen den FIFO erst mit irgendwelchen zufälligen Bits füllen muss, bis die eingestellte Paketlänge erreicht ist. Der Empfang wird dadurch wesentlich schlechter.
Im Bild sieht man das recht deutlich. In der Zeit etwa 10:00 Uhr bis 11:30 hatte ich rfmode Bresser_6in1 eingestellt.

![MsgCount](https://user-images.githubusercontent.com/26026988/229359696-e62c213a-c398-42e8-b291-6e65c5d93ce8.png)

Reply:
Bin mich am einlesen, wie ist das denn mit dem GDO0_CFG.

Das steht ja auf 0x01, demnach müsste folgendes Verhalten vorliegen oder

```
Associated to the RX FIFO: Asserts when RX FIFO is filled at or above the RX FIFO threshold or the end of packet is
reached. De-asserts when the RX FIFO is empty
```

Bedeutet doch, dass der GDO0 erst auf high geht, wenn der FIFO bis oder über den Schwellwert gefüllt wurde. Richtig?
Komisch, dass ich dann auch Signale mit 14 Bit bekomme, wenn der Schwellwert bei 16 Bit steht.

Reply:
"RX FIFO threshold" spielt wahrscheinlich keine Rolle, weil im Register PKTCTRL0 (0x0008) die Bits für "LENGTH_CONFIG" auf "0 : Fixed packet length mode. Length configured in PKTLEN register" eingestellt ist.

Ich habe bewusst diesen Modus gewählt, weil sich "RX FIFO threshold" nicht so fein einstellen lässt. Das lässt sich nur in Stufen von jeweils 4 Byte einstellen.

Reply:
Wenn ich das richtig erfasst habe, dann haben wir Sensoren die 11 Byte senden und welche die 14 Byte senden.

Dieser hier sendet 11 Byte, also drei weniger als die fixe Länge im Register, wird aber empfangen. 
Ich hab PKTLEN bei mir auf 11 gesetzt und in 20 Minuten den Sensor 3x empfangen.  Nicht berauschend oft, aber irgendwie klappt das manchmal mit den 11 Byte.
Mal schauen ob sich was ändert, wenn ich den FIFOTHR  auf 8 Byte stelle

Reply:
Wenn du den so selten empfängst, ist etwas anderes faul. Der Sensor sendet wohl etwa jede Minute.
Ich schätze mal, die Frequenz passt nicht richtig.
Es gibt ein Register, das die Abweichung der Empfangsfrequenz von eingestellten ausgibt:

```
      		int8_t freqErr = cc1101::readReg(0x32, CC1101_READ_BURST); // 0x32 (0xF2): FREQEST – Frequency Offset Estimate from Demodulator
```

Das musst du allerdings umrechnen:

```
    $freqafc = $msg_parts{freqafc};
    ## -128 to 127 (255)
    if ($freqafc > 127) {$freqafc -= 255;}
    $freqafc = int(26000000 / 16384 * $freqafc / 1000);
```

Reply:
Das müsste ich direkt nach dem Empfang auslesen oder? Also uC und Modul anpassen. Richtig?

Reply:
Ja, scheint aber auch so ganz gut zu gehen. Ich habe mal eben immer mal Register 0x32 ausgelesen und dann die Empfangsfrequenz angepasst:

![FrOffset](https://user-images.githubusercontent.com/26026988/229374582-45f1619c-3b74-46bf-a31b-81cb66749ad3.png)

Die Anpassung der Firmware und des SIGNALduino-Modules schlummert hier eigentlich schon seit einem halben Jahr auf der Festplatte bzw. läuft auch schon auf mehreren SIGNALduinos. Es lässt sich dadurch auch eine Art AFC realisieren.
Das Problem ist nur das scharfe Regex, das dann faulty msg auswirft, wenn nicht auch gleichzeitig mit der Firmware das SIGNALduino-Modul aktualisiert wird.

Reply:
Ich habe mir die Ausgabe eingebaut. Naja irgendwas hab ich da nicht richtig gemacht. Es taucht ja nicht auf.
Wenn der Wert für mehr als nur Debug Ausgaben interessant ist, können wir ihn ja dauerhaft einbauen.

Nachdem ich jetzt einen crashkurs mit dem FSK Code hinter mir habe :)

Reply:
Mir kam da auch in den Sinn, dass man erst den FIFO ausliest, wenn der Threshold erreicht wurde.
Dann müsste man eigentlich nachschauen auf welchem Wert die Paketlänge steht. GGf. kurz warten und alle Daten abrufen.

Reply:
> Ich habe mir die Ausgabe eingebaut. Naja irgendwas hab ich da nicht richtig gemacht. Es taucht ja nicht auf. Wenn der Wert für mehr als nur Debug Ausgaben interessant ist, können wir ihn ja dauerhaft einbauen.

Brauchbar ist der Wert auf jeden Fall. Gerade in kritischen Empfangsbedingungen sollte die Empfangsfrequenz möglichst genau eingestellt werden.
Wir haben an die Nachricht eine weitere Variable angehangen "A=", wo der Wert aus Register 0x32 ausgegeben wird:

`sduinoD1_RAWMSG   MN;D=E8377FEBDFEEEFCEFF9C9FBA7F17C88014201110310063604580;R=234;A=-3;
`

Das SIGNALduino-Modul muss allerdings auch umgebaut werden, um diesen Wert dann umgerechnet in kHz auszugeben:

`sduinoD1_FREQAFC   -4`

Reply:
@elektron-bbs 
Ich habe mich nun dazu entschieden length_max zu entfernen, da die maximal bereitgestellte Länge vom gewählten RF Mode abhängt und ich nicht in die Zukunft sehen kann, was noch alles kommt.